README.md

```diff
@@ -0,0 +1,84 @@
+| Global | "assetMBR" | Stores the ASA MBR |
+| Box | "s-" + `auth-addr` | Mapping of signer to open opt-in signature |
+| Box |  "e-" + `auth-addr` | Mapping of signer to open opt-in end time |
+| Box | "s-" + `sha256(signer,sender)` | Mapping of the hash of the signer address plus sender address to address opt-in signature |
+| Box |  "e-" + `sha256(signer,sender)` | Mapping of the hash of the signer address plus sender address to address opt-in end time |
+
+## Rationale
+Box storage is used to store signatures indefinitely and [the verifier lsig](./contracts/verifier_lsig.teal) ensures the signature is always correct.
+
+End time functionality is provided to allow users to revert the effects of signign the logic signature programs without having to rekey. It also let's users open their account for opt-in delegations for a short amount of time without having to remember to manually undo it.
+
+Asset MBR is stored in global storage in case the MBR for asset were to ever change.
+
+## Backwards Compatibility
+N/A
+
+## Test Cases
+Tests written with Jest and algokit for the contract and logic signatures can be seen in [test.ts](./tests/test.ts).
+
+```
+Delegated Opt In App
+    create
+      ✓ creates the app (1427 ms)
+    setSigVerificationAddress
+      ✓ works with valid address (939 ms)
+    setOpenOptInSignature
+      ✓ works with valid signature and lsig (1031 ms)
+    openOptIn
+      ✓ works with valid lsig and method call (973 ms)
+    setOpenOptInEndTime
+      ✓ works with 0xffffffff as the end time (2091 ms)
+      ✓ works with 0 as the end time (1137 ms)
+    setAddressOptInSignature
+      ✓ works with valid signature and lsig (926 ms)
+    addressoptOptIn
+      ✓ works with valid lsig, method call, and sender (914 ms)
+    setAddressOptInEndtime
+      ✓ works with 0xffffffff as the end time (1001 ms)
+      ✓ works with 0 as the end time (1611 ms)
+```
+
+## Reference Implementation
+[delegated_optin_app.algo.ts](./contracts/delegated_optin_app.algo.ts) is the application written in TEALScript.
+
+[DelegatedOptIn.approval.teal](./contracts/artifacts/DelegatedOptIn.approval.teal) is the TEAL compiled from TEALScript.
+
+## Security Considerations
+
+The test cases test proper functionality of all methods, but there has been no extended effort in attempt to break the contract. Most of the functionality in the app and logic signatures is relatively simple, so the chances of unexpected bugs is relatively low.
+
+It should be made clear that signatures are stored mapped to the signer (`auth-addr`) and end times are mapped to the public Algorand address. This means when a signature is made known, every account with the signing account as the `auth-addr` can now be opted in to asset via the delegated logic signature. Every account must set the end times to their desire, because without an end time set the signature will work regardless of the latest timestamp.
```

Comment:
```suggestion
It should be made clear that signatures are stored mapped to the signer (`auth-addr`) and end times are mapped to the public Algorand address. This means when a signature is made known, every account with the signing account as the `auth-addr` can now be opted-in to assets via the delegated logic signature. Every account must set the end times to their desire, because without an end time set the signature will work regardless of the latest timestamp.
```